### PR TITLE
Add a metaclass to fix CUSTOM_STATES bug

### PR DESCRIPTION
**Pull request by adam-incuna**
@meshy review please? :)

### CODE REVIEW

**Comment by meshy** [comment]
tests failed

**Comment by adam-incuna** [comment]
@meshy yeah just saw that, looks like I forgot to update requirements - cheers :)

**Comment by meshy** [comment]
:+1:

File: incuna_auth/models.py [hunk 1]
```diff
@@ -1,6 +1,30 @@
 from django.db import models
+from six import add_metaclass
```
Comment (meshy) [inline incuna_auth/models.py]: this should be `django.six`

Comment (adam-incuna) [inline incuna_auth/models.py]: It failed when I did that.

Comment (adam-incuna) [inline incuna_auth/models.py]: screenshot from 2015-01-15 16 26 17

**Comment by meshy** [comment]
Shit, I was too late, sorry. There's no need to add six to the requirements, as it's bundled with django. Instead, you should import from `django.six`.

**Comment by adam-incuna** [comment]
That didn't work when I did it :/ (see screenshot)

**Comment by meshy** [comment]
Oops, my bad! It's `django.utils.six`

**Comment by adam-incuna** [comment]
@meshy at long last, it passes!

File: incuna_auth/models.py [hunk 2]
```diff
@@ -1,6 +1,36 @@
...
+        # If `attrs` does not contain BASE_ACCESS_STATES, look up the inheritance tree
+        # by iterating over the elements in `bases` to find the first one that has
+        # BASE_ACCESS_STATES on it.
+        if not base_states:
+            get_base_states = lambda b: hasattr(b, base_attr_name)
+            base_states_source = next(filter(get_base_states, bases))
+            base_states = getattr(base_states_source, base_attr_name)
+
+        if base_states:
+            attrs['ACCESS_STATES'] = custom_states + base_states
```
Comment (meshy) [inline incuna_auth/models.py]: Is this `if` required?

File: tests/test_models.py [hunk 3]
```diff
@@ -21,8 +21,15 @@
...
     """Test that AccessStateExtensionMixin does what it's supposed to do."""
 
-    class AccessState(models.AccessStateExtensionMixin):
-        CUSTOM_STATES = (CUSTOM_STATE,)
+    def test_custom_states(self):
+        """Assert that setting the CUSTOM_STATE attribute works properly."""
+        class AccessState(models.AccessStateExtensionMixin):
+            CUSTOM_STATES = (CUSTOM_STATE,)
+
+        access = AccessState
```
Comment (meshy) [inline tests/test_models.py]: I don't think this variable makes things any clearer.

File: incuna_auth/models.py (hunk 2 above)
Comment (adam-incuna) [inline incuna_auth/models.py]: Ah probably not - you said the iteration would throw a `StopIteration`, right?

File: tests/test_models.py (hunk 3 above)
Comment (adam-incuna) [inline tests/test_models.py]: It used to be `AccessState()`.  I'll remove it :)

File: incuna_auth/models.py (hunk 2 above)
Comment (meshy) [inline incuna_auth/models.py]: That's right

**Comment by meshy** [comment]
Looks good `:)` I'm happy when Travis is.

**Comment by adam-incuna** [comment]
Thanks! :)

**Comment by meshy** [comment]
...travis wasn't happy! `:P`

**Comment by adam-incuna** [comment]
I saw a green tick, I tell you :P (it's just more Travis errors.)

**Comment by meshy** [comment]
It failed because the branch was deleted before it started `:P`

**Pull request merged**